### ISSUE TITLE
NaN when using big-endian arrays

### ISSUE THREAD

**Issue by alfiopuglisi** [cupy/cupy#3652]
When a big-endian array is loaded on the GPU using cp.array(), random NaNs appear in the data and calculations will start returning NaN. No errors or warnings are given to the user.

### Conditions
CuPy Version          : 7.6.0
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 9010
CUDA Driver Version   : 10010
CUDA Runtime Version  : 9010
cuBLAS Version        : 9010
cuFFT Version         : 9010
cuRAND Version        : 9010
cuSOLVER Version      : (9, 1, 0)
cuSPARSE Version      : 9010
NVRTC Version         : (9, 1)
cuDNN Build Version   : 7102
cuDNN Version         : 7102
NCCL Build Version    : 2115
NCCL Runtime Version  : (unknown)
CUB Version           : None
cuTENSOR Version      : None

### Code to reproduce
```import cupy as cp
import numpy as np

data = np.arange(1000*1000, dtype='>f4')/1e9
print(' numpy:', type(data), data.shape, data.dtype)
print('   nan:', np.where(np.isnan(data)))
print(' total:', np.sum(data))

arr = cp.array(data)
print('-----')
print('  cupy:' , type(arr), arr.shape, arr.dtype)
print('   nan:' , cp.where(cp.isnan(arr)))
print(' total:', cp.sum(arr))`
```

### Output of the above code:

```
 numpy: <class 'numpy.ndarray'> (1000000,) >f4
   nan: (array([], dtype=int64),)
 total: 499.99963
-----
  cupy: <class 'cupy.core.core.ndarray'> (1000000,) >f4
   nan: (array([   213,    385,    426, ..., 999227, 999242, 999391]),)
 total: nan
```

The numpy array shows no NaNs as expected, while the cupy array on the GPU shows several NaNs and functions like sum() that operate on the whole array return NaN as well.

### Scenarios

A fairly common occurence in a scientific environment is when readings FITS files, which store data big-endian, for example using the astropy.io module:

```
import cupy as cp
from astropy.io import fits

data = fits.getdata(filename)
gpu_data = cp.array(data)        # Results in NaN
```

A workaround is to convert the array to little endian before using it with cupy:

`data = data.astype(np.fits32)`

**Comment by leofang** [comment]
According to the CUDA Programming Guide:
> The NVIDIA GPU architecture uses a little-endian representation.

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#hardware-implementation

**Comment by alfiopuglisi** [comment]
True, but numpy supports little- and big-endianess completely transparently, and I didn't even know that I was trying to load a big-endian array. It took some hours of debugging to figure it out.

I think that in such a situation cupy should either:
- give an error; or
- automatically convert the dtype while moving the array to/from the device

currently, it takes a perfectly valid numpy array, and it loads wrong data on the device. I think that it is a bug.

**Comment by alfiopuglisi** [comment]
Found a similar issue: #2744 ndarray.dot produces wrong values with non-native endian arrays

**Comment by leofang** [comment]
For the record I am no decision maker, so don't take my words too seriously. I think it's long weekend in Japan, so someone from PFN would eventually reply. 

Part of me also agrees it's a bug. We need to signal the incompatibility as it's very hard to debug as you said. However, another part of me wonders how you ended up with a big-endian array (in a real code, not a reproducer) given that most of the mainstream CPU architectures are little endian too (which was why CUDA picks it). You need to do very naughty things to get there, don't you?

**Comment by alfiopuglisi** [comment]
I would never create big-endian arrays on purpose, but it can happen if you use third party libraries that produce numpy arrays, like the astropy example I gave in the original report. 

I triggered the bug just reading files from disk using astropy and loading them on the GPU, three lines of code. Nothing naughty :)  Since FITS files are very common were I work, I now need to wrap the cupy loading code with some checks, otherwise it will happen all the time.

**Comment by leofang** [comment]
> but it can happen if you use third party libraries that produce numpy arrays, like the astropy example I gave in the original report.

Sorry, @alfiopuglisi, I overlooked your astropy example. So I took a quick search, and think they're the ones being naughty 😂 It's been an issue repeating over the years of astropy (which is an impressive package, although I never was a user of it), and they still think it's OK to live with...Wow.

A few refs I found:
- https://github.com/astropy/astropy/issues/1156
- https://github.com/astropy/astropy/issues/4069
- https://github.com/astropy/astropy/pull/3729
- https://github.com/astropy/astropy/pull/4080

You might be able to call `Table.as_array()` as a workaround before CuPy core devs figure out what to do?

**Comment by kmaehashi** [comment]
When converting a NumPy array into a CuPy array we automatically make array contiguous (C by default). So I think it is also reasonable to automatically convert an array into big endian before sending it to GPU.

**Comment by joseph-long** [comment]
Weighing in to report I had a similar issue. (I was trying to use multidimensional fancy indexing to turn images into vectors and back.)

The right behavior here is, IMO, transparent conversion to little-endian in the host-to-device copy. At any rate, this _can't_ be right...

<img width="565" alt="image" src="https://user-images.githubusercontent.com/31168267/95286219-990d7c80-0817-11eb-86c3-7a2e949e83d7.png">

For what it's worth, I don't think Astropy will change, since using the same byte-order as the FITS standard specifies lets them do memory mapped array access easily.

(P.S. Hi @alfiopuglisi ! I'm using CuPy to reduce data from MagAO! Fancy running in to one of its creators here...)

**Comment by alfiopuglisi** [comment]
Here is the workaround that I have implemented in my code: it patches the cupy array and asarray functions to convert the data on-the-fly. It's a hack but it works for me, maybe it will be useful for others too:

```py
def _make_sure_first_arg_is_little_endian(f):
    '''
    Decorator to enforce little endianess with cupy.

    Analyze the first argument and, if it is
    a big-endian array (like the ones returned by astropy.io.fits.getdata),
    convert it to a little-endian one before passing it to the function.

    Can be used to "patch" the cupy routines as follows::
        cupy.array = _make_sure_first_arg_is_little_endian(cupy.array)
        cupy.asarray = _make_sure_first_arg_is_little_endian(cupy.asarray)
    '''
    @wraps(f)
    def wrapper(data, *args, **kwargs):
        import sys
        import numpy

        if isinstance(data, numpy.ndarray):
            # Use names compatible with sys.byteorder
            endianess_map = {
                '>': 'big',
                '<': 'little',
                '=': sys.byteorder,
                '|': 'not relevant'
            }
            endianess = endianess_map[data.dtype.byteorder]
            if endianess == 'big':
               print('Array %s %s converted from big to little endian' %
                     (str(data.dtype), str(data.shape)))
               data = data.astype(data.dtype.newbyteorder('<'))

        return f(data, *args, **kwargs)

    return wrapper


def cupy_patch(cupy):
    '''
    Patch the passed cupy module with our workarounds for FITS files.
    '''
    cupy.array = _make_sure_first_arg_is_little_endian(cupy.array)
    cupy.asarray = _make_sure_first_arg_is_little_endian(cupy.asarray)    

```

PS: Hi @joseph-long! MagAO was a lot of fun, happy to see that you are working with the data!
